imageJPG.ml

```diff
@@ -17,71 +17,373 @@
+  let default_chrominance_quantize_table = [|
+    17; 18; 24; 47; 99; 99; 99; 99;
+    18; 21; 26; 66; 99; 99; 99; 99;
+    24; 26; 56; 99; 99; 99; 99; 99;
+    47; 66; 99; 99; 99; 99; 99; 99;
+    99; 99; 99; 99; 99; 99; 99; 99;
+    99; 99; 99; 99; 99; 99; 99; 99;
+    99; 99; 99; 99; 99; 99; 99; 99;
+    99; 99; 99; 99; 99; 99; 99; 99
+  |]
+
+  (* The JPEG entropy encoder uses a zig-zag order. *)
+  let make_zigzag () =
+    let xy_to_index x y =
+      x + 8*y
+    in
+    let rec loop acc = function
+      (* (X, Y, dir (if any), moved) *)
+      (* Reached the end -> return *)
+      | (7, 7, _) ->
+        acc
+      (* Top/bottom -> go left, then southwest until we reach the edge *)
+      | (_, 0, _) | (_, 7, _) as m ->
+        let (x, y, _) = m in
+        loop (xy_to_index x y :: xy_to_index (x + 1) y :: acc) (x, y - 1, -1)
+      (* Left/right -> go down, then northeast until we reach the edge *)
+      | (0, _, _) | (7, _, _) as m ->
+        let (x, y, _) = m in
+        loop (xy_to_index x y :: xy_to_index x (y - 1) :: acc) (x + 1, y, +1)
+      (* Anything else -> keep going *)
+      | (x, y, dir) ->
+        loop (xy_to_index x y :: acc) (x + dir, y + dir, dir)
+    in
+    Array.of_list (loop [] (0, 0, 0))
+
+  type component = {
+    identifier : int;
+    h_sample_factor : int;
+    v_sample_factor : int;
+    quant_index : int;
+  }
+
+  type sof = {
+    precision : int;
+    height : int;
+    width : int;
+    components : component array;
+  }
+
+  type dqt = {
+    precision : int;
+    index : int;
+    entries : int array;
+  }
+
+  type component_spec = {
+    component_index : int;
+    dc_table_index : int;
+    ac_table_index : int;
+  }
+
+  type sos = {
+    component_specs : component_spec list;
+    first_dct_component : int;
+    last_dct_component : int;
+    approx_high : int;
+    approx_low : int;
+  }
+
   let extensions = ["jpg"; "jpeg"; "jpe"; "jif"; "jfif"; "jfi"]
 
-  let read_marker ich =
-    let ff = chunk_byte ich in
-    if ff <> 0xff then
-      raise (Corrupted_image "Expected marker...");
-    let rec read_first_not_ff ich =
-      let c = chunk_byte ich in
-      if c = 0xff then read_first_not_ff ich else c
+  let string_of_marker c =
+    let mask_char c mask =
+      string_of_int @@ int_of_char c land mask
+    in
+    match c with
+    | '\x01' -> "TEM"
+    | '\xC0' .. '\xC3' | '\xC5' .. '\xC7' | '\xC9' .. '\xCB' | '\xCD' .. '\xCF' as c -> "SOF" ^ mask_char c 0x0F
+    | '\xC4' -> "DHT"
+    | '\xCC' -> "DAC"
+    | '\xD0' .. '\xD7' as c -> "RST" ^ mask_char c 0x07
+    | '\xD8' -> "SOI"
+    | '\xD9' -> "EOI"
+    | '\xDA' -> "SOS"
+    | '\xDB' -> "DQT"
+    | '\xDC' -> "DNL"
+    | '\xDD' -> "DRI"
+    | '\xDE' -> "DHP"
+    | '\xDF' -> "EXP"
+    | '\xE0' .. '\xEF' as c -> "APP" ^ mask_char c 0x0F
+    | '\xF0' .. '\xFD' as c -> "JPG" ^ mask_char c 0x0F
+    | '\xFE' -> "COM"
+    | _ as c -> raise (Corrupted_image (Printf.sprintf "%C is not a valid marker" c))
```

Comment:
Below you have `    | c -> handle_function ich c.[0]`,
could we change this from `| _ as c ->` to just `| c ->` ?
At least I find that easier to read.